File: samples/grpc/iot-service-scala/src/main/scala/iot/IotServer.scala
package iot

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.util.Failure
import scala.util.Success

import akka.actor.typed.ActorSystem
import akka.grpc.scaladsl.ServerReflection
import akka.grpc.scaladsl.ServiceHandler
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse

object IotServer {

  def start(
      interface: String,
      port: Int,
      system: ActorSystem[_],
      registrationService: registration.proto.RegistrationService,
      registrationEventProducerService: PartialFunction[
        HttpRequest,
        Future[HttpResponse]],
      pushedTemperatureEventsHandler: PartialFunction[
        HttpRequest,
        Future[HttpResponse]],
      sensorTwinService: temperature.proto.SensorTwinService): Unit = {
    implicit val sys: ActorSystem[_] = system
    implicit val ec: ExecutionContext =
      system.executionContext

    val service: HttpRequest => Future[HttpResponse] =
      ServiceHandler.concatOrNotFound(
        registrationEventProducerService,
        registration.proto.RegistrationServiceHandler
          .partial(registrationService),
        pushedTemperatureEventsHandler,
        temperature.proto.SensorTwinServiceHandler.partial(sensorTwinService),
        // ServerReflection enabled to support grpcurl without import-path and proto parameters
        ServerReflection.partial(
          List(
            registration.proto.RegistrationService,
            temperature.proto.SensorTwinService)))

    val bound =
      Http()
        .newServerAt(interface, port)
        .bind(service)
        .map(_.addToCoordinatedShutdown(3.seconds))

    bound.onComplete {
      case Success(binding) =>
        val address = binding.localAddress
        system.log.info(
          "IoT Service online at gRPC server {}:{}",
          address.getHostString,
          address.getPort)
      case Failure(ex) =>
        system.log.error("Failed to bind gRPC endpoint, terminating system", ex)
        system.terminate()
    }
  }

}

File: samples/grpc/iot-service-java/src/main/java/iot/IotServer.java
package iot;

import akka.actor.typed.ActorSystem;
import akka.grpc.javadsl.ServerReflection;
import akka.grpc.javadsl.ServiceHandler;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.japi.function.Function;
import iot.registration.proto.RegistrationService;
import iot.registration.proto.RegistrationServiceHandlerFactory;
import iot.temperature.proto.SensorTwinService;
import iot.temperature.proto.SensorTwinServiceHandlerFactory;
import java.time.Duration;
import java.util.Arrays;
import java.util.concurrent.CompletionStage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class IotServer {

  private static final Logger logger = LoggerFactory.getLogger(IotServer.class);

  public static void start(
      ActorSystem<?> system,
      String host,
      int port,
      RegistrationService registrationService,
      Function<HttpRequest, CompletionStage<HttpResponse>> registrationEventProducerService,
      SensorTwinService sensorTwinService,
      Function<HttpRequest, CompletionStage<HttpResponse>> pushedTemperatureEventsHandler) {

    @SuppressWarnings("unchecked")
    var service =
        ServiceHandler.concatOrNotFound(
            RegistrationServiceHandlerFactory.create(registrationService, system),
            SensorTwinServiceHandlerFactory.create(sensorTwinService, system),
            registrationEventProducerService,
            pushedTemperatureEventsHandler,
            ServerReflection.create(
                Arrays.asList(RegistrationService.description, SensorTwinService.description),
                system));

    var bound = Http.get(system).newServerAt(host, port).bind(service);

    bound.whenComplete(
        (binding, error) -> {
          if (error == null) {
            logger.info(
                "IoT Service online at gRPC server {}:{}",
                binding.localAddress().getHostString(),
                binding.localAddress().getPort());
            binding.addToCoordinatedShutdown(Duration.ofSeconds(3), system);
          } else {
            logger.error("Failed to bind gRPC endpoint, terminating system", error);
            system.terminate();
          }
        });
  }
}
